cmd/network/setup_linux.go

```diff
@@ -0,0 +1,206 @@
+	vsockDialPort      = 6656
+	signaturePhrase    = "github.com/rancher-sandbox/rancher-desktop-networking"
+	readySignal        = "READY"
+)
+
+func main() {
+	flag.BoolVar(&debug, "debug", false, "enable additional debugging")
+	flag.StringVar(&vmSwitchPath, "vm-switch-path", "", "the path to the vm-switch binary that will run in a new namespace")
+	flag.StringVar(&vmSwitchLogFile, "vm-switch-logfile", "", "path to the logfile for vm-switch process")
+	flag.StringVar(&unshareArg, "unshare-arg", "", "the command argument to pass to the unshare program")
+	flag.StringVar(&logFile, "logfile", "/var/log/network-setup.log", "path to the logfile for network setup process")
+	flag.Parse()
+
+	if err := log.SetOutputFile(logFile, logrus.StandardLogger()); err != nil {
+		logrus.Fatalf("setting logger's output file failed: %v", err)
+	}
+
+	if debug {
+		logrus.SetLevel(logrus.DebugLevel)
+	}
+
+	if vmSwitchPath == "" {
+		logrus.Fatal("path to the vm-switch process must be provided")
+	}
+
+	if unshareArg == "" {
+		logrus.Fatal("unshare program arg must be provided")
+	}
+
+	// listenForHandshake blocks until a successful handshake estabilished
+	listenForHandshake()
+
+	logrus.Debugf("attempting to connect to the host on CID: %v and Port: %d", vsock.CIDHost, vsockDialPort)
+	vsockConn, err := vsock.Dial(vsock.CIDHost, vsockDialPort)
+	if err != nil {
+		logrus.Fatal(err)
+	}
+	logrus.Debugf("successful connection to host on CID:%v and Port: %d: connection: %+v", vsock.CIDHost, vsockDialPort, vsockConn)
+
+	// setup network namespace
+	ns, err := configureNamespace()
+	if err != nil {
+		logrus.Fatal(err)
+	}
+
+	if err := unshareCmd(ns, unshareArg); err != nil {
+		logrus.Fatal(err)
+	}
+
+	// Start the vm-switch process in the new namespace,
+	// the reason behind this approach is to avoid switching
+	// back and forth between namespaces.
+	// This is due to the limitaion in the golang's runtime
+	// all the sub process could potentially start in the original
+	// namespace.
+	args := []string{
+		fmt.Sprintf("-n/proc/%d/fd/%d", os.Getpid(), ns),
+		"-F",
+		vmSwitchPath,
+	}
+	if vmSwitchLogFile != "" {
+		args = append(args, "-logfile", vmSwitchLogFile)
+	}
+	if debug {
+		args = append(args, "-debug")
+	}
+	vmSwitchCmd := exec.Command(nsenter, args...)
+
+	connFile, err := vsockConn.File()
+	if err != nil {
+		logrus.Fatal(err)
+	}
+	// pass in the vsock connection as a FD to the
+	// vm-switch process in the newely created namespace
+	vmSwitchCmd.ExtraFiles = []*os.File{connFile}
+	if err := vmSwitchCmd.Start(); err != nil {
+		logrus.Fatalf("could not start the vm-switch process: %v", err)
+	}
+	logrus.Infof("successfully started the vm-switch running with a PID: %v", vmSwitchCmd.Process.Pid)
+
+	errGroup := errgroup.Group{}
+
+	errGroup.Go(func() error {
+		if err := vmSwitchCmd.Wait(); err != nil {
+			return fmt.Errorf("vm-switch exited with error: %w", err)
+		}
+		return nil
+	})
+
+	if err = errGroup.Wait(); err != nil {
```

Comment:
It's kind of weird to have an `errgroup` that just has one thing in it… why not just get rid of that whole thing and just do `vmSwitchCmd.Wait()` directly?